core/services/src/main/java/org/exoplatform/services/attachments/service/AttachmentServiceImpl.java

```diff
@@ -235,46 +260,86 @@ public void deleteAttachmentItemById(long userIdentityId,
-      try {
-        for (AttachmentContextEntity attachmentContextEntity : attachmentsContextEntity) {
-          Attachment attachment = EntityBuilder.fromAttachmentNode(repositoryService,
-                                                                   documentService,
-                                                                   workspace,
-                                                                   session,
-                                                                   attachmentContextEntity.getAttachmentId());
-          boolean canView = canView(userIdentityId, entityType, entityId);
-          boolean canDelete = canDelete(userIdentityId, entityType, entityId);
-          Permission attachmentACL = new Permission(canView, canDelete);
-          attachment.setAcl(attachmentACL);
-          attachments.add(attachment);
-        }
-      } finally {
-        if (session != null) {
-          session.logout();
-        }
+    List<Attachment> entityAttachments = attachmentStorage.getAttachmentsByEntity(userIdentityId, entityId, entityType);
+    if (!entityAttachments.isEmpty()) {
+      boolean canView = canView(userIdentityId, entityType, entityId);
+      boolean canDelete = canDelete(userIdentityId, entityType, entityId);
+      boolean canEdit = canEdit(userIdentityId, entityType, entityId);
+      Permission attachmentACL = new Permission(canView, canDelete, canEdit);
+      entityAttachments.forEach(attachment -> attachment.setAcl(attachmentACL));
+    }
+    return entityAttachments;
+  }
+
+  @Override
+  public Attachment getAttachmentById(String entityType,
+                                      long entityId,
+                                      String attachmentId,
+                                      long userIdentityId) throws ObjectNotFoundException {
+    Attachment attachment = new Attachment();
+    Session session = null;
+    try {
+      session = Utils.getSession(sessionProviderService, repositoryService);
+      attachment = EntityBuilder.fromAttachmentNode(repositoryService,
+                                                    documentService,
+                                                    linkManager,
+                                                    Utils.getCurrentWorkspace(repositoryService),
+                                                    session,
+                                                    attachmentId);
+
+      boolean canView = canView(userIdentityId, entityType, entityId);
+      boolean canDelete = canDelete(userIdentityId, entityType, entityId);
+      boolean canEdit = canEdit(userIdentityId, entityType, entityId);
+      Permission attachmentACL = new Permission(canView, canDelete, canEdit);
+
+      attachment.setAcl(attachmentACL);
+    } catch (Exception e) {
+      throw new ObjectNotFoundException("Can't convert attachment JCR node with id " + attachmentId + " to entity");
```

Comment:
not compatible exceptions, can't transform all `Exception` => `ObjectNotFoundException`